File: lib/utils/cloudflare-bypass.js
/**
 * Cloudflare Bypass Utility - Node.js wrapper för cloudscraper (Python)
 *
 * Använder cloudscraper via Python subprocess för att:
 * - Hämta giltiga cookies som passerar Cloudflare
 * - Få user-agent som fungerar med sessionen
 *
 * @module cloudflare-bypass
 */

const { spawn } = require('child_process');
const path = require('path');

// Sökväg till Python-scriptet
const PYTHON_SCRIPT = path.join(__dirname, 'cloudflare-bypass.py');

// Sökväg till venv Python
const VENV_PYTHON = path.join(__dirname, '../../.venv/bin/python3');

/**
 * Kör Python cloudflare-bypass scriptet
 * @param {string[]} args - Argument till scriptet
 * @returns {Promise<Object>} - JSON-resultat från scriptet
 */
function runPythonScript(args) {
    return new Promise((resolve, reject) => {
        // Försök med venv python först, fallback till system python
        const pythonPaths = [VENV_PYTHON, 'python3', '/usr/bin/python3'];

        let currentIndex = 0;

        function tryNextPython() {
            if (currentIndex >= pythonPaths.length) {
                reject(new Error('Kunde inte hitta Python med cloudscraper installerat'));
                return;
            }

            const pythonPath = pythonPaths[currentIndex];
            currentIndex++;

            const proc = spawn(pythonPath, [PYTHON_SCRIPT, ...args], {
                timeout: 60000,
                env: { ...process.env, PYTHONUNBUFFERED: '1' }
            });

            let stdout = '';
            let stderr = '';

            proc.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            proc.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            proc.on('error', (err) => {
                // Försök nästa Python-path
                tryNextPython();
            });

            proc.on('close', (code) => {
                if (code !== 0 && stderr.includes('No module named')) {
                    // cloudscraper ej installerat, försök nästa
                    tryNextPython();
                    return;
                }

                try {
                    const result = JSON.parse(stdout);
                    resolve(result);
                } catch (e) {
                    reject(new Error(`Kunde inte parsa output: ${stdout}\nStderr: ${stderr}`));
                }
            });
        }

        tryNextPython();
    });
}

/**
 * Hämta cookies som passerar Cloudflare för en URL
 * @param {string} url - URL att bypassa
 * @param {Object} options - Alternativ
 * @param {boolean} options.turnstile - Använd Turnstile-specifik lösning
 * @param {boolean} options.html - Inkludera HTML i svaret
 * @returns {Promise<{success: boolean, cookies: Array, user_agent: string, error?: string}>}
 */
async function bypassCloudflare(url, options = {}) {
    const args = [url];

    if (options.turnstile) {
        args.push('--turnstile');
    }
    if (options.html) {
        args.push('--html');
    }

    console.log(`[CloudflareBypass] Försöker bypassa: ${url}`);

    try {
        const result = await runPythonScript(args);

        if (result.success) {
            console.log(`[CloudflareBypass] Lyckades! Fick ${result.cookies?.length || 0} cookies`);
        } else {
            console.log(`[CloudflareBypass] Misslyckades: ${result.error}`);
        }

        return result;
    } catch (error) {
        console.error(`[CloudflareBypass] Error: ${error.message}`);
        return {
            success: false,
            error: error.message,
            error_type: 'wrapper_error'
        };
    }
}

/**
 * Applicera cookies från bypass till en Puppeteer-sida
 * @param {Object} page - Puppeteer page object
 * @param {Array} cookies - Cookies från bypassCloudflare
 * @param {string} domain - Domän att sätta cookies för
 */
async function applyCookiesToPage(page, cookies, domain = 'www.ratsit.se') {
    if (!cookies || !Array.isArray(cookies)) {
        console.log('[CloudflareBypass] Inga cookies att applicera');
        return;
    }

    const puppeteerCookies = cookies.map(c => ({
        name: c.name,
        value: c.value,
        domain: c.domain || domain,
        path: c.path || '/',
        secure: c.secure || false,
        httpOnly: c.httpOnly || false,
        sameSite: c.sameSite || 'Lax'
    }));

    await page.setCookie(...puppeteerCookies);
    console.log(`[CloudflareBypass] Applicerade ${puppeteerCookies.length} cookies till sidan`);
}

/**
 * Hämta en sida via cloudscraper och returnera HTML
 * Användbart för sidor som blockerar Puppeteer helt
 * @param {string} url - URL att hämta
 * @returns {Promise<{success: boolean, html?: string, cookies: Array}>}
 */
async function fetchPageWithBypass(url) {
    return bypassCloudflare(url, { html: true, turnstile: true });
}

module.exports = {
    bypassCloudflare,
    applyCookiesToPage,
    fetchPageWithBypass
};

File: lib/utils/cloudflare-bypass.py
#!/usr/bin/env python3
"""
Cloudflare Bypass Utility - Använder cloudscraper för att bypassa Cloudflare-skydd

Användning:
    python3 cloudflare-bypass.py <url> [--cookies] [--html]

Output (JSON):
    {
        "success": true,
        "cookies": [...],
        "html": "...",
        "user_agent": "..."
    }

Används av Node.js scrapers för att få giltiga cookies/session.
"""

import sys
import json
import cloudscraper

def bypass_cloudflare(url, return_html=False):
    """
    Bypassa Cloudflare-skydd och returnera cookies + user-agent
    """
    try:
        # Skapa scraper med browser-emulering
        scraper = cloudscraper.create_scraper(
            browser={
                'browser': 'chrome',
                'platform': 'darwin',  # macOS
                'desktop': True
            },
            delay=10  # Vänta på JS-challenges
        )

        # Gör request för att lösa Cloudflare challenge
        response = scraper.get(url, timeout=30)

        # Extrahera cookies
        cookies = []
        for cookie in scraper.cookies:
            cookies.append({
                'name': cookie.name,
                'value': cookie.value,
                'domain': cookie.domain,
                'path': cookie.path,
                'secure': cookie.secure,
                'httpOnly': cookie.has_nonstandard_attr('HttpOnly'),
                'sameSite': 'Lax'
            })

        result = {
            'success': True,
            'status_code': response.status_code,
            'cookies': cookies,
            'user_agent': scraper.headers.get('User-Agent', ''),
            'url': response.url
        }

        if return_html:
            result['html'] = response.text[:50000]  # Begränsa storlek

        return result

    except cloudscraper.exceptions.CloudflareChallengeError as e:
        return {
            'success': False,
            'error': f'Cloudflare challenge failed: {str(e)}',
            'error_type': 'cloudflare_challenge'
        }
    except Exception as e:
        return {
            'success': False,
            'error': str(e),
            'error_type': type(e).__name__
        }

def solve_turnstile(url):
    """
    Försök lösa Cloudflare Turnstile specifikt
    """
    try:
        scraper = cloudscraper.create_scraper(
            browser={
                'browser': 'chrome',
                'platform': 'darwin',
                'desktop': True
            },
            delay=15,  # Längre delay för Turnstile
            captcha={
                'provider': 'return_response'  # Returnera challenge-info
            }
        )

        response = scraper.get(url, timeout=45)

        cookies = []
        for cookie in scraper.cookies:
            cookies.append({
                'name': cookie.name,
                'value': cookie.value,
                'domain': cookie.domain,
                'path': cookie.path,
                'secure': cookie.secure
            })

        return {
            'success': response.status_code == 200,
            'status_code': response.status_code,
            'cookies': cookies,
            'user_agent': scraper.headers.get('User-Agent', ''),
            'cf_clearance': next((c['value'] for c in cookies if c['name'] == 'cf_clearance'), None)
        }

    except Exception as e:
        return {
            'success': False,
            'error': str(e),
            'error_type': type(e).__name__
        }

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print(json.dumps({
            'success': False,
            'error': 'Usage: python3 cloudflare-bypass.py <url> [--cookies] [--html] [--turnstile]'
        }))
        sys.exit(1)

    url = sys.argv[1]
    return_html = '--html' in sys.argv
    use_turnstile = '--turnstile' in sys.argv

    if use_turnstile:
        result = solve_turnstile(url)
    else:
        result = bypass_cloudflare(url, return_html)

    print(json.dumps(result, ensure_ascii=False))
